skyd/skyd.go

```diff
@@ -1,16 +1,137 @@
 package skyd
 
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/SkynetLabs/skyd/node/api/client"
+	"gitlab.com/SkynetLabs/skyd/skymodules"
+)
+
+var (
+	// skylinksCache is a local cache of the list of skylinks pinned by the
+	// local skyd
+	skylinksCache pinnedSkylinksCache
+	// skylinksCacheDuration defines the duration of skylinksCache
+	skylinksCacheDuration = time.Hour
+)
+
 type (
 	// Client allows us to call the local skyd instance.
 	Client struct {
+		staticClient *client.Client
+	}
+	// ClientInterface describes the interface exposed by Client, so we can
+	// mock it for testing.
+	ClientInterface interface {
+		Pin(skylink string) error
+		PinnedSkylinks() (skylinks []string, err error)
+		Unpin(skylink string) error
+	}
+
+	// renterDirCache is a simple cache of the renter's directory information,
+	// so we don't need to fetch that for each skylink we potentially want to
+	// pin/unpin.
+	pinnedSkylinksCache struct {
+		Skylinks   []string
+		Expiration time.Time
+		mu         sync.Mutex
 	}
 )
 
-/*
-TODO
- - env vars for talking to skyd - host, port, apipassword, etc.
- - methods for calling skyd:
-	- list all pinned skylinks
-	- pin a skylink
-	- unpin a skylink
-*/
+// NewClient creates a new skyd client.
+func NewClient(host, port, password string) *Client {
+	opts := client.Options{
+		Address:       fmt.Sprintf("%s:%s", host, port),
+		Password:      password,
+		UserAgent:     "Sia-Agent",
+		CheckRedirect: nil,
+	}
+	return &Client{
+		staticClient: client.New(opts),
+	}
+}
+
+// Pin instructs the local skyd to pin the given skylink.
+func (c *Client) Pin(skylink string) error {
+	pinned, err := c.isPinned(skylink)
+	if err != nil {
+		return err
+	}
+	if pinned {
+		// The skylink is already locally pinned, nothing to do.
+		return nil
+	}
+	sp, err := skymodules.NewSiaPath(skylink)
+	if err != nil {
+		return errors.AddContext(err, "failed to create siapath")
+	}
+	spp := skymodules.SkyfilePinParameters{
+		SiaPath: sp,
+	}
+	return c.staticClient.SkynetSkylinkPinPost(skylink, spp)
+}
+
+// PinnedSkylinks returns the list of skylinks pinned by the local skyd.
+func (c *Client) PinnedSkylinks() (skylinks []string, err error) {
+	skylinksCache.mu.Lock()
+	defer skylinksCache.mu.Unlock()
+	// Check whether the cache is still valid and return it if so.
+	if skylinksCache.Expiration.After(time.Now().UTC()) {
+		return skylinksCache.Skylinks, nil
+	}
+	// The cache is not valid, fetch the data from skyd.
+	sls := make([]string, 0)
+	dirsToWalk := []skymodules.SiaPath{skymodules.SkynetFolder}
+	for len(dirsToWalk) > 0 {
```

Comment:
Yes, we did. I created an issue for it: https://gitlab.com/SkynetLabs/skyd/-/issues/328
But since it doesn't exist at this point, I kept this implementation, rather than blocking the entire Pinner project.
I see this as a nice F/U rather than a blocking issue that needs to be resolved before we can put this in prod.

Reply:
ok makes sense, just confirming.

also follow ups are by definition not blocking. Just should be done. :-)